src/metabase/models/serialization/base.clj

```diff
@@ -472,3 +498,136 @@
+            (get collections (:collection_id entity)) ;; This can be nil, but that's fine - that's the root collection.
+            [(lower-plural model) (storage-leaf-file-name id label)])]
+    #_(clojure.pprint/pprint {:entity entity :path result})
+    result))
+
+(defmulti storage-path
+  "Computes the complete storage path for a given entity.
+  `(storage-path entity ctx)`
+  Dispatches on the model name, eg. \"Dashboard\".
+
+  Returns a list of strings giving the path, with the final entry being the file name with no extension.
+
+  The default implementation works for entities which are:
+  - Part of the regular (not snippet) collections tree, per a :collection_id field; and
+  - Stored as `foos/1234abc_slug.extension` underneath their collection
+  eg. Cards, Dashboards, Timelines
+
+  The default logic is captured by [[storage-default-collection-path]] so it can be reused."
+  (fn [entity _] (ingested-model entity)))
+
+(defmethod storage-path :default [entity ctx]
+  (storage-default-collection-path entity ctx))
+
+(defn storage-base-context
+  "Creates the basic context for storage. This is a map with a single entry: `:collections` is a map from collection ID
+  to the path of collections."
+  []
+  (let [colls      (db/select ['Collection :id :entity_id :location :slug])
+        coll-names (into {} (for [{:keys [id entity_id slug]} colls]
+                              [(str id) (storage-leaf-file-name entity_id slug)]))
+        coll->path (into {} (for [{:keys [entity_id id location]} colls
+                                  :let [parents (rest (str/split location #"/"))]]
+                              [entity_id (map coll-names (concat parents [(str id)]))]))]
+    {:collections coll->path}))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                               Ingestion                                                        |
+;;; +----------------------------------------------------------------------------------------------------------------+
+(defonce ^:private ingest-path-matchers (atom {}))
+
+(defn ingest-path
+  "Transforms a file path (as a sequence of strings with no file extension on the last part) into a `:serdes/meta`
+  hierarchy, or nil if nothing matches.
+
+  The set of matchers is maintained in an atom, and each model should call [[register-ingestion-path]] to add its
+  matcher.
+
+  Note that the input format is the same as the return value of [[storage-path]]."
+  [path]
+  (first (keep #(% path) (vals @ingest-path-matchers))))
+
+(defn register-ingestion-path
+  "Registers the matcher for the given model. Expects the model name to be a string.
+  The matcher is a function from a path (see [[ingest-path]] for the format) to a `:serdes/meta` hierarchy, or nil
+  if there's no match."
+  [model-name matcher]
+  (swap! ingest-path-matchers assoc model-name matcher)
+  ['register-ingestion-path model-name])
```

Comment:
A comment explaining why it's returning this could be nice. It's not obvious TBH. I'm guessing maybe you're using it for test purposes or something

Reply:
It makes it show up nicely in the REPL when you eval the buffer or the `register-ingestion-path!` call.
Commented accordingly.